### ISSUE TITLE
Turn Go warnings into errors

### ISSUE THREAD

**Issue by adrianludwin** [kubernetes-sigs/hierarchical-namespaces#146]
/good-first-issue

Let's try to prevent having to do #138 all over again.

**Comment by nobbs** [comment]
Hey!

I've added staticcheck (https://staticcheck.io/) as a step in the `Makefile`. This seems to be one of the de-facto static code analysis tools for go. The change is minimal (see f52ebfcee433a7e6d31ad2d07fa185d08ef3d308); it will install staticcheck and run it against the same files as `go vet` either with `make staticcheck` or during `build`, `docker-build`, `docker-push-multi`:

One remaining problem right now is, that there's still a bunch of warnings left (for 9c08a710995b703e5ed67a269a8076662d00f7b1), these should be fixed before this is going to be used as a default step.

```
❯ make staticcheck 
staticcheck ./api/... ./cmd/... ./internal/...
internal/anchor/reconciler.go:245:13: error strings should not be capitalized (ST1005)
internal/anchor/reconciler.go:295:13: error strings should not be capitalized (ST1005)
internal/anchor/reconciler.go:358:22: func (*Reconciler).deleteInstance is unused (U1000)
internal/anchor/validator.go:58:10: webhooks.Deny is deprecated: Use DenyFromAPIError instead; please don't add new callers.  (SA1019)
internal/anchor/validator.go:103:11: webhooks.Deny is deprecated: Use DenyFromAPIError instead; please don't add new callers.  (SA1019)
internal/anchor/validator.go:108:11: webhooks.Deny is deprecated: Use DenyFromAPIError instead; please don't add new callers.  (SA1019)
internal/anchor/validator.go:117:12: webhooks.Deny is deprecated: Use DenyFromAPIError instead; please don't add new callers.  (SA1019)
internal/anchor/validator.go:126:11: webhooks.Deny is deprecated: Use DenyFromAPIError instead; please don't add new callers.  (SA1019)
internal/config/namespace.go:101:11: error strings should not be capitalized (ST1005)
internal/config/namespace.go:104:11: error strings should not be capitalized (ST1005)
internal/forest/namespace.go:151:5: should omit comparison to bool constant, can be simplified to ns.allowCascadingDeletion (S1002)
internal/hierarchyconfig/validator.go:14:2: package "k8s.io/apimachinery/pkg/api/errors" is being imported more than once (ST1019)
        internal/hierarchyconfig/validator.go:15:2: other import of "k8s.io/apimachinery/pkg/api/errors"
internal/hierarchyconfig/validator.go:95:10: webhooks.Deny is deprecated: Use DenyFromAPIError instead; please don't add new callers.  (SA1019)
internal/hierarchyconfig/validator.go:128:10: webhooks.Deny is deprecated: Use DenyFromAPIError instead; please don't add new callers.  (SA1019)
internal/hierarchyconfig/validator.go:132:10: webhooks.Deny is deprecated: Use DenyFromAPIError instead; please don't add new callers.  (SA1019)
internal/hierarchyconfig/validator.go:185:10: webhooks.Deny is deprecated: Use DenyFromAPIError instead; please don't add new callers.  (SA1019)
internal/hierarchyconfig/validator.go:193:10: webhooks.Deny is deprecated: Use DenyFromAPIError instead; please don't add new callers.  (SA1019)
internal/hierarchyconfig/validator.go:203:10: webhooks.Deny is deprecated: Use DenyFromAPIError instead; please don't add new callers.  (SA1019)
internal/hierarchyconfig/validator.go:213:10: webhooks.Deny is deprecated: Use DenyFromAPIError instead; please don't add new callers.  (SA1019)
internal/hierarchyconfig/validator.go:218:10: webhooks.Deny is deprecated: Use DenyFromAPIError instead; please don't add new callers.  (SA1019)
internal/hierarchyconfig/validator.go:227:10: webhooks.Deny is deprecated: Use DenyFromAPIError instead; please don't add new callers.  (SA1019)
internal/hierarchyconfig/validator.go:235:10: webhooks.Deny is deprecated: Use DenyFromAPIError instead; please don't add new callers.  (SA1019)
internal/hierarchyconfig/validator.go:407:12: webhooks.Deny is deprecated: Use DenyFromAPIError instead; please don't add new callers.  (SA1019)
internal/hierarchyconfig/validator.go:412:12: webhooks.Deny is deprecated: Use DenyFromAPIError instead; please don't add new callers.  (SA1019)
internal/hierarchyconfig/validator.go:419:12: webhooks.Deny is deprecated: Use DenyFromAPIError instead; please don't add new callers.  (SA1019)
internal/hierarchyconfig/validator.go:423:12: webhooks.Deny is deprecated: Use DenyFromAPIError instead; please don't add new callers.  (SA1019)
internal/hncconfig/reconciler_test.go:457:6: func updateTypeConfig is unused (U1000)
internal/hncconfig/reconciler_test.go:486:6: func unsetTypeConfig is unused (U1000)
internal/hncconfig/reconciler_test.go:583:16: should use fmt.Errorf(...) instead of errors.New(fmt.Sprintf(...)) (S1028)
internal/hncconfig/reconciler_test.go:587:14: should use fmt.Errorf(...) instead of errors.New(fmt.Sprintf(...)) (S1028)
internal/hncconfig/reconciler_test.go:604:17: should use fmt.Errorf(...) instead of errors.New(fmt.Sprintf(...)) (S1028)
internal/hncconfig/reconciler_test.go:621:16: should use fmt.Errorf(...) instead of errors.New(fmt.Sprintf(...)) (S1028)
internal/hncconfig/reconciler_test.go:625:14: should use fmt.Errorf(...) instead of errors.New(fmt.Sprintf(...)) (S1028)
internal/hncconfig/validator.go:52:11: webhooks.Deny is deprecated: Use DenyFromAPIError instead; please don't add new callers.  (SA1019)
internal/hncconfig/validator.go:65:10: webhooks.Deny is deprecated: Use DenyFromAPIError instead; please don't add new callers.  (SA1019)
internal/hncconfig/validator.go:142:11: webhooks.Deny is deprecated: Use DenyFromAPIError instead; please don't add new callers.  (SA1019)
internal/integtest/helpers.go:9:2: should not use dot imports (ST1001)
internal/integtest/helpers.go:10:2: should not use dot imports (ST1001)
internal/integtest/helpers.go:84:3: unnecessary assignment to the blank identifier (S1005)
internal/integtest/helpers.go:92:3: unnecessary assignment to the blank identifier (S1005)
internal/integtest/helpers.go:380:2: unnecessary assignment to the blank identifier (S1005)
internal/integtest/setup.go:24:2: should not use dot imports (ST1001)
internal/integtest/setup.go:25:2: should not use dot imports (ST1001)
internal/namespace/validator.go:56:10: webhooks.Deny is deprecated: Use DenyFromAPIError instead; please don't add new callers.  (SA1019)
internal/namespace/validator.go:140:11: webhooks.Deny is deprecated: Use DenyFromAPIError instead; please don't add new callers.  (SA1019)
internal/namespace/validator.go:148:12: webhooks.Deny is deprecated: Use DenyFromAPIError instead; please don't add new callers.  (SA1019)
internal/namespace/validator.go:176:10: webhooks.Deny is deprecated: Use DenyFromAPIError instead; please don't add new callers.  (SA1019)
internal/namespace/validator.go:187:10: webhooks.Deny is deprecated: Use DenyFromAPIError instead; please don't add new callers.  (SA1019)
internal/namespace/validator.go:204:11: webhooks.Deny is deprecated: Use DenyFromAPIError instead; please don't add new callers.  (SA1019)
internal/namespace/validator.go:219:10: webhooks.Deny is deprecated: Use DenyFromAPIError instead; please don't add new callers.  (SA1019)
internal/namespace/validator.go:237:10: webhooks.Deny is deprecated: Use DenyFromAPIError instead; please don't add new callers.  (SA1019)
internal/namespace/validator.go:250:11: webhooks.Deny is deprecated: Use DenyFromAPIError instead; please don't add new callers.  (SA1019)
internal/objects/validator.go:84:11: webhooks.Deny is deprecated: Use DenyFromAPIError instead; please don't add new callers.  (SA1019)
internal/objects/validator.go:94:11: webhooks.Deny is deprecated: Use DenyFromAPIError instead; please don't add new callers.  (SA1019)
internal/objects/validator.go:128:11: webhooks.Deny is deprecated: Use DenyFromAPIError instead; please don't add new callers.  (SA1019)
internal/objects/validator.go:134:11: webhooks.Deny is deprecated: Use DenyFromAPIError instead; please don't add new callers.  (SA1019)
internal/objects/validator.go:138:11: webhooks.Deny is deprecated: Use DenyFromAPIError instead; please don't add new callers.  (SA1019)
internal/objects/validator.go:141:11: webhooks.Deny is deprecated: Use DenyFromAPIError instead; please don't add new callers.  (SA1019)
internal/objects/validator.go:144:11: webhooks.Deny is deprecated: Use DenyFromAPIError instead; please don't add new callers.  (SA1019)
internal/objects/validator.go:150:11: webhooks.Deny is deprecated: Use DenyFromAPIError instead; please don't add new callers.  (SA1019)
internal/objects/validator.go:251:10: webhooks.Deny is deprecated: Use DenyFromAPIError instead; please don't add new callers.  (SA1019)
internal/objects/validator.go:259:11: webhooks.Deny is deprecated: Use DenyFromAPIError instead; please don't add new callers.  (SA1019)
internal/objects/validator.go:263:11: webhooks.Deny is deprecated: Use DenyFromAPIError instead; please don't add new callers.  (SA1019)
internal/objects/validator.go:273:11: webhooks.Deny is deprecated: Use DenyFromAPIError instead; please don't add new callers.  (SA1019)
internal/objects/validator.go:280:11: webhooks.Deny is deprecated: Use DenyFromAPIError instead; please don't add new callers.  (SA1019)
internal/objects/validator.go:286:11: webhooks.Deny is deprecated: Use DenyFromAPIError instead; please don't add new callers.  (SA1019)
internal/objects/validator.go:295:9: webhooks.Deny is deprecated: Use DenyFromAPIError instead; please don't add new callers.  (SA1019)
internal/stats/stats.go:139:31: argument total is overwritten before first use (SA4009)
        internal/stats/stats.go:144:3: assignment to total
internal/stats/stats.go:148:7: should omit comparison to bool constant, can be simplified to !working (S1002)
internal/stats/stats.go:156:7: should omit comparison to bool constant, can be simplified to working (S1002)
make: *** [Makefile:195: staticcheck] Error 1
```

**Comment by erikgb** [comment]
I have submitted https://github.com/kubernetes-sigs/hierarchical-namespaces/pull/137 to fix most of the callers to the deprecated func. PTAL!

**Comment by nobbs** [comment]
@erikgb Here's the output against the state of #137 - looks way better now:

```
❯ staticcheck ./api/... ./cmd/... ./internal/...
internal/anchor/reconciler.go:245:13: error strings should not be capitalized (ST1005)
internal/anchor/reconciler.go:295:13: error strings should not be capitalized (ST1005)
internal/anchor/reconciler.go:358:22: func (*Reconciler).deleteInstance is unused (U1000)
internal/config/namespace.go:101:11: error strings should not be capitalized (ST1005)
internal/config/namespace.go:104:11: error strings should not be capitalized (ST1005)
internal/forest/namespace.go:151:5: should omit comparison to bool constant, can be simplified to ns.allowCascadingDeletion (S1002)
internal/hncconfig/reconciler_test.go:457:6: func updateTypeConfig is unused (U1000)
internal/hncconfig/reconciler_test.go:486:6: func unsetTypeConfig is unused (U1000)
internal/hncconfig/reconciler_test.go:583:16: should use fmt.Errorf(...) instead of errors.New(fmt.Sprintf(...)) (S1028)
internal/hncconfig/reconciler_test.go:587:14: should use fmt.Errorf(...) instead of errors.New(fmt.Sprintf(...)) (S1028)
internal/hncconfig/reconciler_test.go:604:17: should use fmt.Errorf(...) instead of errors.New(fmt.Sprintf(...)) (S1028)
internal/hncconfig/reconciler_test.go:621:16: should use fmt.Errorf(...) instead of errors.New(fmt.Sprintf(...)) (S1028)
internal/hncconfig/reconciler_test.go:625:14: should use fmt.Errorf(...) instead of errors.New(fmt.Sprintf(...)) (S1028)
internal/integtest/helpers.go:9:2: should not use dot imports (ST1001)
internal/integtest/helpers.go:10:2: should not use dot imports (ST1001)
internal/integtest/helpers.go:84:3: unnecessary assignment to the blank identifier (S1005)
internal/integtest/helpers.go:92:3: unnecessary assignment to the blank identifier (S1005)
internal/integtest/helpers.go:380:2: unnecessary assignment to the blank identifier (S1005)
internal/integtest/setup.go:24:2: should not use dot imports (ST1001)
internal/integtest/setup.go:25:2: should not use dot imports (ST1001)
internal/objects/validator.go:149:11: webhooks.Deny is deprecated: Use one of the explicit deny reason funcs instead; please don't add new callers.  (SA1019)
internal/objects/validator.go:250:10: webhooks.Deny is deprecated: Use one of the explicit deny reason funcs instead; please don't add new callers.  (SA1019)
internal/objects/validator.go:262:11: webhooks.Deny is deprecated: Use one of the explicit deny reason funcs instead; please don't add new callers.  (SA1019)
internal/objects/validator.go:272:11: webhooks.Deny is deprecated: Use one of the explicit deny reason funcs instead; please don't add new callers.  (SA1019)
internal/objects/validator.go:279:11: webhooks.Deny is deprecated: Use one of the explicit deny reason funcs instead; please don't add new callers.  (SA1019)
internal/objects/validator.go:285:11: webhooks.Deny is deprecated: Use one of the explicit deny reason funcs instead; please don't add new callers.  (SA1019)
internal/stats/stats.go:139:31: argument total is overwritten before first use (SA4009)
        internal/stats/stats.go:144:3: assignment to total
internal/stats/stats.go:148:7: should omit comparison to bool constant, can be simplified to !working (S1002)
internal/stats/stats.go:156:7: should omit comparison to bool constant, can be simplified to working (S1002)
```

**Comment by erikgb** [comment]
Nice! I am planning to fix the deprecated callers. Feel free to fix the remaining warnings in a dedicated PR!

**Comment by nobbs** [comment]
Yeah, will do in the next few days :)

**Comment by adrianludwin** [comment]
Fixed by #169 

/close
